src/can.rs

```diff
@@ -1,657 +1,94 @@
-    ) -> Can {
+    /// Note: this does not actually initialize the CAN bus. You will need to first call [`bxcan::Can::new`] and  set the bus configuration and filters before the peripheral can be enabled. See the CAN example, for a more thorough example of the full setup process.
+    pub fn new(can: pac::CAN, tx: TX, rx: RX, apb1: &mut APB1) -> Self {
         apb1.enr().modify(|_, w| w.canen().enabled());
-        can.mcr.modify(|_, w| w.sleep().clear_bit());
-        can.mcr.modify(|_, w| w.inrq().set_bit());
-
-        // Wait for INAK to confirm we have entered initialization mode
-        while !can.msr.read().inak().bit_is_set() {}
-
-        let CanOpts {
-            brp,
-            sjw,
-            ts1,
-            ts2,
-            lbkm,
-        } = opts;
-        can.btr.modify(|_, w| unsafe {
-            w.brp()
-                .bits(brp)
-                .sjw()
-                .bits(sjw)
-                .ts1()
-                .bits(ts1)
-                .ts2()
-                .bits(ts2)
-                .lbkm()
-                .variant(lbkm)
-        });
-
-        // Leave initialization mode by clearing INRQ and switch to normal mode
-        can.mcr.modify(|_, w| w.inrq().clear_bit());
-        // Wait for INAK again to confirm the mode switch
-        while !can.msr.read().inak().bit_is_clear() {}
+        apb1.rstr().modify(|_, w| w.canrst().set_bit());
+        apb1.rstr().modify(|_, w| w.canrst().clear_bit());
 
         Can {
-            can,
-            _rx: rx,
+            _can: can,
             _tx: tx,
+            _rx: rx,
         }
     }
-    /// Initialize the CAN Peripheral using default options from `CanOpts::default()`
-    pub fn new(
-        can: pac::CAN,
-        rx: gpioa::PA11<AF9<PushPull>>,
-        tx: gpioa::PA12<AF9<PushPull>>,
-        apb1: &mut APB1,
-    ) -> Self {
-        Can::new_with_opts(can, rx, tx, apb1, CanOpts::default())
-    }
-
-    /// Enable CAN event interrupts for `Event`
-    pub fn listen(&mut self, event: Event) {
-        self.can.ier.modify(|_, w| match event {
-            Event::Fifo0Fmp => w.fmpie0().set_bit(),
-            Event::Fifo0Full => w.ffie0().set_bit(),
-            Event::Fifo0Ovr => w.fovie0().set_bit(),
-            Event::Txe => w.tmeie().set_bit(),
-            _ => unimplemented!(),
-        });
-    }
-
-    /// Split the CAN peripheral into a transmitter and associated FIFOs.
-    pub fn split(self) -> (CanTransmitter, CanFifo, CanFifo) {
-        let fifo0 = CanFifo { idx: 0 };
-        let fifo1 = CanFifo { idx: 1 };
-
-        let transmitter = CanTransmitter {
-            _can: self.can,
-            _rx: self._rx,
-            _tx: self._tx,
-        };
-
-        (transmitter, fifo0, fifo1)
-    }
-
-    /// Release owned peripherals
-    pub fn free(
-        self,
-    ) -> (
-        pac::CAN,
-        gpioa::PA11<AF9<PushPull>>,
-        gpioa::PA12<AF9<PushPull>>,
-    ) {
-        (self.can, self._rx, self._tx)
-    }
-}
```

Comment:
Hm. I see, the `bxcan` crate does consume this peripheral and is not able to free it. But I would still keep this function and submit a PR to the `bxcan` crate, that it also supports freeing the peripheral again.

Reply:
[bxcan 0.5.1](https://github.com/stm32-rs/bxcan/blob/master/CHANGELOG.md#051---2021-05-15) now has a `Can::free` method, so we can use it here. So bxcan instance would return `fn free(bxcan::Can<Self>) -> Self`, where Self itself implements `pub fn free(self) -> (pac::CAN, gpioa::PA11<...>, gpioa::PA12 <...>)`. 

This would mean, we have to call free twisce to get back the gpios: `let (can, pa11, pa12) = can.free().free();` but I think, it is not too bad.